File: backend/src/main/java/com/skewwhiffy/auraltester/notation/model/scale/Scale.java
package com.skewwhiffy.auraltester.notation.model.scale;

import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote;
import com.skewwhiffy.auraltester.notation.model.note.NoteSequence;
import lombok.val;

import java.text.MessageFormat;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public record Scale(AbsoluteNote lowestNote, ScaleType scaleType, ScaleDirection direction)
        implements NoteSequence {
    public String getDisplayName() {
        return MessageFormat.format(
                "{0} {1}",
                lowestNote.note().getDisplayString(),
                scaleType.displayName()
        );
    }

    @Override
    public List<AbsoluteNote> getNotes() {
        val notes = scaleType
                .getIntervals(direction)
                .stream()
                .map(lowestNote::apply);
        return (switch (direction) {
            case ASCENDING -> notes.toList();
            case DESCENDING -> notes.collect(
                    Collectors.collectingAndThen(
                            Collectors.toList(),
                            it -> {
                                Collections.reverse(it);
                                return it;
                            }
                    )
            );
        });
    }
      /*
  fun abc(key: Key): String = notes.joinToString("") { it.abc(key) }
     */
}

File: backend/src/main/kotlin/com/skewwhiffy/auraltester/notation/model/scale/Scale.kt
package com.skewwhiffy.auraltester.notation.model.scale

import com.skewwhiffy.auraltester.notation.model.note.AbsoluteNote
import com.skewwhiffy.auraltester.notation.model.note.NoteSequence
import com.skewwhiffy.auraltester.notation.model.note.displayString

data class Scale(val lowestNote: AbsoluteNote, val scaleType: ScaleType, val direction: ScaleDirection) : NoteSequence {
    val displayName
        get() = "${lowestNote.note.displayString} ${scaleType.displayName}"

    override val notes: List<AbsoluteNote>
        get() {
            val notes = scaleType
                .getIntervals(direction)
                .map(lowestNote::apply)
            return when (direction) {
                ScaleDirection.ASCENDING -> notes
                ScaleDirection.DESCENDING -> notes.reversed()
            }
        }
}
